ldelf/ta_elf_rel.c

```diff
@@ -42,11 +42,14 @@ static bool __resolve_sym(struct ta_elf *elf, unsigned int bind,
 	if (!st_name)
 		return false;
 	if (st_name > elf->dynstr_size)
-		err(TEE_ERROR_BAD_FORMAT, "Symbol out of range");
+		err(TEE_ERROR_BAD_FORMAT, "Symbol name out of range");
 
 	if (strcmp(name, elf->dynstr + st_name))
 		return false;
 
+	if (st_value > (elf->max_addr - elf->load_addr))
+		err(TEE_ERROR_BAD_FORMAT, "Symbol location out of range");
```

Comment:
Hmmmm... not always OK. Depends on the symbol type. For "usual" symbols that we may encounter at load time (`STT_OBJECT` and `STT_FUNC` for instance), we do expect an address. But, consider the `STT_TLS` type for instance. The [spec](https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-79797.html) says:

> STT_TLS
> The symbol specifies a thread-local storage entity. When defined, this symbol gives the assigned offset for the symbol, not the actual address.

I happen to know this well, because I'm currently working on C++ support and in order to be able to use the `arm-linux-gnueabihf-` toolchain, I had to implement a relocation type which will typically lookup such symbols.

That being said, the fact that we're adding `elf->load_addr` assumes that this function can only be used for symbols that represent addresses...

How about passing the symbol type to the function? Then the proper check can be done and `elf->load_addr` can be added only when applicable. Which will integrate well with my future patches :wink:

Reply:
Makes sense, I'll add and the check symbol type.